kvserver: add setting to use expiration-based leases

erikgrinaker opened the pull request:
**kvserver: rename `Replica.requiresExpiringLease()` to `Expiration`**

Expiration leases is the common term for these.

Epic: none
Release note: None

**kvserver: add setting to use expiration-based leases**

This patch adds the experimental cluster setting `kv.expiration_leases_only.enabled`, which uses expiration-based leases for all ranges. The setting is marked as experimental because, while we believe the system will function correctly, it has performance implications that need to be mapped out and optimized.

Expiration-based leases are compelling because they are much more robust than epoch-based leases, and better handle failure modes such as partial/asymmetric network partitions, disk stalls, deadlocks, etc. They require a Raft roundtrip to extend a lease, which ensures that the lease must be functional, while epoch leases only require an RPC request to the liveness leaseholder regardless of whether the lease actually works.

Except for the meta and liveness ranges, expiration leases are only extended when a request is processed in the last half of the lease interval (i.e. in the last 3 seconds of the 6 second lease). Otherwise, we allow the lease to expire. This reduces the cost of idle ranges in large clusters, since we avoid the periodic lease extension writes for every range, and can let the ranges quiesce as usual. However, it incurs a synchronous lease acquisition on the next request to the range.

Because expiration leases incur one Raft write per range per lease extension, as well as a lease acquisition for idle ranges, they currently come with significant performance overhead. In TPC-E experiments at 100.000 customers with various transaction types, p50 latencies increased 5-70%, p99 latencies increased 20-80%, and pMax latencies increased 0-1000%. A kv95 workload on 10.000 ranges with active leases showed a throughput reduction of about 10%, most likely due to the ~3.000 lease extension writes per second.

When the setting is changed, leases are asynchronously switched to the appropriate type (either expiration-based or epoch-based) via the replicate queue. This can take up to the `ScanInterval` to complete, 10 minutes by default (more on nodes with large numbers or ranges).

Resolves #93903.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/94457)
<!-- Reviewable:end -->

andrewbaptist reviewed (changes requested):
This is really neat! 

I'm not sure about the cost of the TODO comment. Does this get called on every request to a range?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NKZn0GF15P5odlHseW_:b-rou2bd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NKZnjJh2zxN3X5nIRcl:-NKZnjJh2zxN3X5nIRcm:bid9ur) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf32e6a166e7da5d1cae8066db20a6b6918ce997/pkg/kv/kvserver/replica_range_lease.go#L1433)):*
> ```Go
> 		r.requiresExpiringLeaseRLocked()
> 
> 	if isExpirationLease != shouldUseExpirationLease {
> ```

Is this logic correct - if the current lease is an expiration and the setting is unset (using epoch) - this will return true. But I think that is incorrect. I might just be misreading this.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> I'm not sure about the cost of the TODO comment. Does this get called on every request to a range?

Yes, it's called on every request. I was running some rough benchmarks with an earlier prototype that called this during request processing and Raft ticks, and it showed up prominently on CPU profiles (about 5%). I'm going to run more benchmarks before merging the final version, so this is mostly a reminder to myself to see if it still shows up in profiles.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NKZq7Ur8jAhCAP34C6Z:btlthun)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NKZnjJh2zxN3X5nIRcl:-NKZpgsa1FVtTu2utN4b:b-dhx41u) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf32e6a166e7da5d1cae8066db20a6b6918ce997/pkg/kv/kvserver/replica_range_lease.go#L1433)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

Is this logic correct - if the current lease is an expiration and the setting is unset (using epoch) - this will return true. But I think that is incorrect. I might just be misreading this.
</blockquote></details>

That's correct, this will cause the lease to be asynchronously renewed as an epoch lease when the setting is turned off. Otherwise, the lease will remain an expiration lease until the leaseholder fails to extend it, which may takes a very long time. The intent is for the setting change to take effect across all ranges within a reasonable time (a minute or so).


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NKZnjJh2zxN3X5nIRcl:-NKZsoKL5BRrI9N3swTx:br674f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf32e6a166e7da5d1cae8066db20a6b6918ce997/pkg/kv/kvserver/replica_range_lease.go#L1433)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

That's correct, this will cause the lease to be asynchronously renewed as an epoch lease when the setting is turned off. Otherwise, the lease will remain an expiration lease until the leaseholder fails to extend it, which may takes a very long time. The intent is for the setting change to take effect across all ranges within a reasonable time (a minute or so).
</blockquote></details>

The other option (and maybe this is what you meant) is to allow the expiration lease to expire when the setting is turned off, and reacquire it as an epoch lease afterwards. However, this will affect request latencies, so we instead renew it asynchronously as soon as possible while the previous lease is still valid.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (approved):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 1433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NKZnjJh2zxN3X5nIRcl:-NKZttW13P7hNPZPipxy:biqs7wo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf32e6a166e7da5d1cae8066db20a6b6918ce997/pkg/kv/kvserver/replica_range_lease.go#L1433)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The other option (and maybe this is what you meant) is to allow the expiration lease to expire when the setting is turned off, and reacquire it as an epoch lease afterwards. However, this will affect request latencies, so we instead renew it asynchronously as soon as possible while the previous lease is still valid.
</blockquote></details>

OK - I understand now, I originally thought that if this returned true it would renew as an expiration lease, but the caller of this checks what to do and handles this correctly. Thanks for the clarification.


<!-- Sent from Reviewable.io -->

erikgrinaker commented:
Seems like replicas aren't properly quiescing here. On an idle 3-node 8-core cluster with 100.000 ranges, we're only quiescing 1/3 of replicas, even though there are only about 500 valid expiration leases. This in turn causes the nodes to run at ~80% CPU when idle, just to tick and heartbeat the replicas.

I initially expected this to be because we can't quiesce ranges with invalid leases, and we didn't get around to quiescing them before the leases expired, but logspy showed that the majority of replicas failed to quiesce because there was outstanding replication quota. There were occasional lease requests, but otherwise no other proposals on these ranges, so it's still unclear why there's outstanding quota, but I'll look into it.

This is all to say that further work is required here regardless of which approach we choose: if we allow idle ranges to quiesce then we need to make sure they properly do so, and if we instead eagerly renew the leases then we need to optimize ticking/stepping of unquiesced ranges.

erikgrinaker commented:
I think the outstanding proposal quota thing is likely a red herring, caused by this condition which will claim pending proposal quota on followers (where `proposalQuota` is `nil`):

https://github.com/cockroachdb/cockroach/blob/14ce32433630f22f4042cdee58e5140550059376/pkg/kv/kvserver/replica_raft.go#L522-L525

It seems like this condition should return `false` instead. We have another condition which will check that the replica is a leader:

https://github.com/cockroachdb/cockroach/blob/14ce32433630f22f4042cdee58e5140550059376/pkg/kv/kvserver/replica_raft_quiesce.go#L321-L326

However, given that we can't quiesce on followers, this should really affect quiescence either way.

erikgrinaker commented:
The problem was indeed that we didn't allow quiescing ranges with invalid leases. I added in a condition to allow quiescing with expired expiration-based leases, and we now properly quiesce idle ranges such that the CPU usage on an idle cluster with 100.000 ranges is about the same with the setting enabled and disabled (8% vs 6%).

erikgrinaker commented:
I ran some quick 3-node 8-CPU kv95 benchmarks with 192 concurrency to look at large-scale changes in behavior/performance. At 10.000 ranges, enabling the setting had a significant impact on performance (~20% throughput reduction). This appeared to mostly be due to the additional writes needed for lease extensions, with a doubling of IOPS (2k to 4k per node) and write bandwidth (8 MB/s to 16MB/s). At 1.000 ranges, however, the workload impact was negligible due to the lower lease extension load. So the additional cost appears to mostly be due to the lease extension writes when we allow idle ranges to quiesce. With a read-only workload, I could run 10.000 ranges without a significant impact to the workload, likely because we had more IO headroom.

We'll need to run some more rigorous benchmarks, but I think this shows that the approach here (without eager lease extension and allowing quiescence) is viable. This is compelling because it avoids having to optimize idle unquiesced ranges and come up with a more sophisticated scheme that can eagerly renew thousands of leases per second. Interested to get your thoughts on this before taking this further.

erikgrinaker commented:
@tbg points out that this will result in high latencies for table scans, e.g. `count(*)`, across idle ranges because we're acquiring the leases sequentially. A quick test with 50k empty ranges took 4 seconds with active epoch leases, and 43 seconds with expired extension leases.

erikgrinaker committed:
kvserver: add setting to use expiration-based leases

This patch adds the experimental cluster setting
`kv.expiration_leases_only.enabled`, which uses expiration-based leases
for all ranges. The setting is marked as experimental because, while we
believe the system will function correctly, it has performance
implications that need to be mapped out and optimized.

Expiration-based leases are compelling because they are much more robust
than epoch-based leases, and better handle failure modes such as
partial/asymmetric network partitions, disk stalls, deadlocks, etc. They
require a Raft roundtrip to extend a lease, which ensures that the lease
must be functional, while epoch leases only require an RPC request to
the liveness leaseholder regardless of whether the lease actually works.

Except for the meta and liveness ranges, expiration leases are only
extended when a request is processed in the last half of the lease
interval (i.e. in the last 3 seconds of the 6 second lease). Otherwise,
we allow the lease to expire. This reduces the cost of idle ranges in
large clusters, since we avoid the periodic lease extension writes for
every range, and can let the ranges quiesce as usual. However, it incurs
a synchronous lease acquisition on the next request to the range.

Because expiration leases incur one Raft write per range per lease
extension, as well as a lease acquisition for idle ranges, they
currently come with significant performance overhead. In TPC-E
experiments at 100.000 customers with various transaction types, p50
latencies increased 5-70%, p99 latencies increased 20-80%, and pMax
latencies increased 0-1000%. A kv95 workload on 10.000 ranges with
active leases showed a throughput reduction of about 10%, most likely
due to the ~3.000 lease extension writes per second.

When the setting is changed, leases are asynchronously switched to the
appropriate type (either expiration-based or epoch-based) via the
replicate queue. This can take up to the `ScanInterval` to complete, 10
minutes by default (more on nodes with large numbers or ranges).

Epic: none
Release note: None

erikgrinaker commented:
This should be ready for a final review now.

andrewbaptist reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> - as we discussed yesterday, to avoid slamming the cluster when this setting changed, it would likely be better to have the replicate queue initiate this change. As this is tagged as "experimental" it is probably OK. 

> Reviewed 2 of 10 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NOyhAjgCNBCt8SwEz4f:b-m0dhum)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 84 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NOyfCCVE72l9hMc_Cby:-NOyfCCVE72l9hMc_Cbz:buxcy4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7273be46b3b36b0d4b3aaf4ff41c4e9cc1d9924/pkg/kv/kvserver/replica_range_lease.go#L84)):*
> ```Go
> 	false,
> )
> ```

nit: Potentially add to the description something like: "changing this setting will immediately change all leases in the system to either epoch or expiration based on the setting. With many ranges, this may cause a short availability outage."


<!-- Sent from Reviewable.io -->

erikgrinaker committed:
kvserver: rename `Replica.requireExpiringLease()` to `Expiration`

Expiration leases is the common term for these.

Epic: none
Release note: None

erikgrinaker reviewed (commented):
> to avoid slamming the cluster when this setting changed, it would likely be better to have the replicate queue initiate this change.

This was a good idea, thank you. In addition to being gentler on the cluster, it significantly simplified the PR and avoided changes in RPC hot paths. Would appreciate another look at the updated PR.

Further down the line we should move the responsibility for lease maintenance to a separate scheduler, I opened #98433 for this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NQFmUNl2kzR1z-jtBS_:b-m1rd43)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
FWIW, here's a lease graph when splitting 10.000 ranges, enabling the setting, and then disabling the setting.

The reason for the abrupt drop in epoch leases around 12:57 appeared to be that rebalancing to a 4th cluster node completed (`roachprod` creates 4-node clusters by default), which cleared out a replicate queue backlog.

<img width="955" alt="Screenshot 2023-03-11 at 14 30 06" src="https://user-images.githubusercontent.com/644420/224487526-424a49a1-0d14-4e28-98aa-dfa10c9118e6.png">

andrewbaptist reviewed (changes requested):
Just one comment related to the use of the `requiresExpirationLeaseRLocked` vs `shouldUseExpirationLeaseRLocked` in two places. 

> Reviewed 7 of 10 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NQf1jIU9poh04o6N4J1:b-s17h46)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @erikgrinaker and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 779 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NQf-4huCoZpP6qlpsVZ:-NQf-4huCoZpP6qlpsV_:b-e44k5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8db1e9722e9f02d1e2e58045db984ee2e3de6e7/pkg/kv/kvserver/replica_range_lease.go#L779)):*
> ```Go
> // epoch-based leases, but may temporarily use expiration based leases during
> // lease transfers.
> func (r *Replica) requiresExpirationLeaseRLocked() bool {
> ```

This is a little hard to follow the cases where code should use `requiresExpirationLeaseRLocked` vs `shouldUseExpirationLeaseRLocked`. The only "real" difference seems to be in `leasePostApplyLocked`. For simpler usage, it might be better to have this method be renamed to `requireActiveLeaseRenewal` and change the next method to `useExpirationLeaseRLocked`. The two other places where this method are used seem incorrect (`maybeCampaignOnWakeLocked` and `shouldCampaignOnRedirect`). It seems like we want expiration leases to campaign in both of these cases. 


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 779 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NQf-4huCoZpP6qlpsVZ:-NQf3XBYDaw8J60xNzBa:b-lhyfni) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8db1e9722e9f02d1e2e58045db984ee2e3de6e7/pkg/kv/kvserver/replica_range_lease.go#L779)):*
<details><summary><i>Previously, andrewbaptist (Andrew Baptist) wrote…</i></summary><blockquote>

This is a little hard to follow the cases where code should use `requiresExpirationLeaseRLocked` vs `shouldUseExpirationLeaseRLocked`. The only "real" difference seems to be in `leasePostApplyLocked`. For simpler usage, it might be better to have this method be renamed to `requireActiveLeaseRenewal` and change the next method to `useExpirationLeaseRLocked`. The two other places where this method are used seem incorrect (`maybeCampaignOnWakeLocked` and `shouldCampaignOnRedirect`). It seems like we want expiration leases to campaign in both of these cases. 
</blockquote></details>

The campaign cases also need `requiresExpirationLeaseRLocked`, it's used to avoid a circular dependency on liveness here:

https://github.com/cockroachdb/cockroach/blob/81a9f1d08ac544ba52047382d3b487e2a659fbe8/pkg/kv/kvserver/replica_raft.go#L2000-L2004

https://github.com/cockroachdb/cockroach/blob/81a9f1d08ac544ba52047382d3b487e2a659fbe8/pkg/kv/kvserver/replica_raft.go#L2079-L2089

`requiresExpirationLeaseRLocked` specifies that a range can't, under _any_ circumstances, use an epoch lease because it causes a circular dependency on liveness and stuff breaks. It requires an expiration lease. I feel like that's clearer than `requireActiveLeaseRenewal` -- in particular because all leases require active renewal, but epoch leases do so via node heartbeats.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 779 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NQf-4huCoZpP6qlpsVZ:-NQf5kx_AAE8dmaaZzuT:b-llk9sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8db1e9722e9f02d1e2e58045db984ee2e3de6e7/pkg/kv/kvserver/replica_range_lease.go#L779)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The campaign cases also need `requiresExpirationLeaseRLocked`, it's used to avoid a circular dependency on liveness here:

https://github.com/cockroachdb/cockroach/blob/a64df6f255974392cdd265e52200e06e1ed700a2/pkg/kv/kvserver/replica_raft.go#L2000-L2004

https://github.com/cockroachdb/cockroach/blob/81a9f1d08ac544ba52047382d3b487e2a659fbe8/pkg/kv/kvserver/replica_raft.go#L2079-L2089

`requiresExpirationLeaseRLocked` specifies that a range can't, under _any_ circumstances, use an epoch lease because it causes a circular dependency on liveness and stuff breaks. It requires an expiration lease. I feel like that's clearer than `requireActiveLeaseRenewal` -- in particular because all leases require active renewal, but epoch leases do so via node heartbeats.
</blockquote></details>

Btw, the `leasePostApplyLocked` case for `requiresExpirationLeaseRLocked` actually goes away once we implement the lease scheduler. See e.g. https://github.com/cockroachdb/cockroach/pull/98455/commits/0593c451c41c242df90e04bea29d82038b9d2aad#diff-488a090afc4b6eaf56cd6d13b347bac67cb3313ce11c49df9ee8cd95fd73b3e8L385.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 779 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NQf-4huCoZpP6qlpsVZ:-NQfAqHo8V7c96yl5sfB:b-fkt4x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8db1e9722e9f02d1e2e58045db984ee2e3de6e7/pkg/kv/kvserver/replica_range_lease.go#L779)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Btw, the `leasePostApplyLocked` case for `requiresExpirationLeaseRLocked` actually goes away once we implement the lease scheduler. See e.g. https://github.com/cockroachdb/cockroach/pull/98455/commits/0593c451c41c242df90e04bea29d82038b9d2aad#diff-488a090afc4b6eaf56cd6d13b347bac67cb3313ce11c49df9ee8cd95fd73b3e8L385.
</blockquote></details>

Nevermind, I reread the campaign logic and you're right -- we should always bail out if the range should be using an expiration lease, since this logic is only relevant for epoch leases. I'll update the PR. Thanks for catching this.

I think I have a mild preference for just using the existing naming for now, and we'll remove `requiresExpirationLeaseRLocked` entirely when we implement the lease scheduler.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andrewbaptist and @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_range_lease.go` line 779 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-NQf-4huCoZpP6qlpsVZ:-NQfHXSR5npS9IWNb4wb:b-pf5znw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8db1e9722e9f02d1e2e58045db984ee2e3de6e7/pkg/kv/kvserver/replica_range_lease.go#L779)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Nevermind, I reread the campaign logic and you're right -- we should always bail out if the range should be using an expiration lease, since this logic is only relevant for epoch leases. I'll update the PR. Thanks for catching this.

I think I have a mild preference for just using the existing naming for now, and we'll remove `requiresExpirationLeaseRLocked` entirely when we implement the lease scheduler.
</blockquote></details>

I realized that we do want to check `requiresExpirationLease` in `shouldCampaignOnWake`, because it allows us to more quickly detect a dead leader and call an election when unquiescing a range that uses an expiration-based lease. But I changed `shouldCampaignOnLeaseRequestRedirect` to use `shouldUseExpirationLease`, because the logic only applies to epoch leases. I've updated the code and comments, PTAL.


<!-- Sent from Reviewable.io -->

andrewbaptist reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/94457#-:-NQh5OUI2IiCL8yj3q0p:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

andrewbaptist commented:
Thanks for the quick turnaround on the changes! I'm glad we are getting this in.

erikgrinaker commented:
TFTR!

bors r+

erikgrinaker commented:
bors retry

erikgrinaker commented:
bors retry
